modin/core/io/column_stores/parquet_dispatcher.py

```diff
@@ -15,13 +15,309 @@
+        ----------
+        path : Pathlike
+            Path to dataset.
+        partition_ids : list
+            Array with references to the partitions data.
+        index_columns : list
+            List of index columns specified by pandas metadata.
+
+        Returns
+        -------
+        index : pandas.Index
+            Index of resulting Modin DataFrame.
+        needs_index_sync : bool
+            Whether the partition indices need to be synced with frame
+            index because there's no index column, or at least one
+            index column is a RangeIndex.
+        """
+        from pyarrow.parquet import read_table
+
+        range_index = True
+        column_names_to_read = []
+        for column in index_columns:
+            # According to https://arrow.apache.org/docs/python/generated/pyarrow.Schema.html,
+            # only RangeIndex will be stored as metadata. Otherwise, the default behavior is
+            # to store the index as a column.
+            if isinstance(column, str):
+                column_names_to_read.append(column)
+                range_index = False
+            elif column["name"] is not None:
+                column_names_to_read.append(column["name"])
+
+        # For the second check, let us consider the case where we have an empty dataframe,
+        # that has a valid index.
+        if range_index or (len(partition_ids) == 0 and len(column_names_to_read) != 0):
+            complete_index = (
+                read_table(path, columns=column_names_to_read).to_pandas().index
+            )
+        # Empty DataFrame case
+        elif len(partition_ids) == 0:
+            return [], False
+        else:
+            index_ids = [part_id[0][1] for part_id in partition_ids if len(part_id) > 0]
+            index_objs = cls.materialize(index_ids)
+            complete_index = index_objs[0].append(index_objs[1:])
+        return complete_index, len(index_columns) == 0 or any(
+            not isinstance(c, str) for c in index_columns
+        )
+
+    @classmethod
+    def build_query_compiler(cls, path, columns, index_columns, **kwargs):
+        """
+        Build query compiler from deployed tasks outputs.
+
+        Parameters
+        ----------
+        path : str, path object or file-like object
+            Path to the file to read.
+        columns : list
+            List of columns that should be read from file.
+        index_columns : list
+            List of index columns specified by pandas metadata.
+        **kwargs : dict
+            Parameters of deploying read_* function.
+
+        Returns
+        -------
+        new_query_compiler : BaseQueryCompiler
+            Query compiler with imported data for further processing.
+        """
+        col_partitions, column_widths = cls.build_columns(columns)
+        partition_ids = cls.call_deploy(path, col_partitions, **kwargs)
+        index, needs_index_sync = cls.build_index(path, partition_ids, index_columns)
```

Comment:
```suggestion
        index, sync_index = cls.build_index(path, partition_ids, index_columns)
```
Maybe just so?

Also, can we right away get `row_lengths` from `build_index` as it is done in the parent in order for us not to run additional `ray.get()` below in `part.length()`?

Reply:
Yes we can get the `row_lengths` for some cases.

Reply:
I think it does matter because we need to call ray.get/client.gather (these are blocking operations) again. To avoid this we could return `row_lengths` from `build_index`. This will also conform the parent behavior.

Reply:
The blocking call shouldn't matter for performance because we pass the length to the partition class constructor in `build_partition`, so the partition should have already cached that length.

Reply:
IIUC we pass the length as a `ray.ObjectRef` to the partition class constructor in `build_partition`. Next we call `ray.get` (let's consider Ray for instance) in `build_index` to get index objs of the partitions. After that we again call `ray.get` for every partition to get its length. I think that (calling `ray.get` twice in the main thread) incurs some perf and memory implications. If you could check this, it would be great. I am not against doing this as part of a separate issue.

Reply:
It's a good point that we're passing an object ref instead of the integer width to the partition constructor in `build_partition`. Still, if we're blocking on the index in `build_index`, the same remote function call should also compute length, so the length should be ready and the `ray.get` should be cheap. But like you, I don't know what is the cost of `ray.get` for a small object that's already ready in the object store. I filed #4743 for further work.